File: Java/Easy/20-ValidParentheses.java
// Description: https://leetcode.com/problems/valid-parentheses/description/

/*
    Approach: Stack with HashMap

    Use a Stack to track opening brackets encountered during iteration.
    Store bracket pairs in a HashMap for O(1) matching lookups.
    For each opening bracket, push it onto the stack.
    For each closing bracket, check if stack is non-empty and top matches.
    If mismatch or stack empty on closing bracket, return false.
    After processing, stack should be empty for valid parentheses.

    Time complexity: O(n) - Single pass through the string.
    Space complexity: O(n) - Stack stores at most n/2 opening brackets.
*/

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

// Solution
class Solution {
    public boolean isValid(String s) {
        Stack<Character> openedBrackets = new Stack<>();
        Map<Character, Character> bracketMap = new HashMap<>();
        bracketMap.put(')', '(');
        bracketMap.put(']', '[');
        bracketMap.put('}', '{');

        for (char c : s.toCharArray()) {
            if (bracketMap.containsKey(c)) { // closed bracket case
                if (openedBrackets.isEmpty() || openedBrackets.peek() != bracketMap.get(c))
                    return false;
                openedBrackets.pop();
            } else openedBrackets.push(c);
        }

        return openedBrackets.isEmpty();
    }
}

// Test
class Test {
    public static void main(String[] args) {
        Solution solution = new Solution();

        // Example 1: s = "()"
        System.out.println("Output: " + solution.isValid("()"));

        // Example 2: s = "()[]{}"
        System.out.println("Output: " + solution.isValid("()[]{}"));

        // Example 3: s = "(]"
        System.out.println("Output: " + solution.isValid("(]"));

        // Example 4: s = "([])"
        System.out.println("Output: " + solution.isValid("([])"));

        // Example 5: s = "([)]"
        System.out.println("Output: " + solution.isValid("([)]"));
    }
}

File: C++/Easy/20-ValidParentheses.cc
// Description: https://leetcode.com/problems/valid-parentheses/description/

/*
    Approach: Stack with Hash Map

    Use a stack to track opening brackets encountered during iteration.
    Store bracket pairs in a hash map for O(1) matching lookups.
    For each opening bracket, push it onto the stack.
    For each closing bracket, check if stack is non-empty and top matches.
    If mismatch or stack empty on closing bracket, return false.
    After processing, stack should be empty for valid parentheses.

    Time complexity: O(n) - Single pass through the string.
    Space complexity: O(n) - Stack stores at most n/2 opening brackets.
*/

#include <iostream>
#include <stack>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> openedBrackets;
        unordered_map<char, char> bracketMap = {{')', '('}, {']', '['}, {'}', '{'}};

        for (char c : s) {
            if (bracketMap.find(c) != bracketMap.end()) {
                if (openedBrackets.empty() || openedBrackets.top() != bracketMap[c]) return false;
                openedBrackets.pop();
            } else
                openedBrackets.push(c);
        }

        return openedBrackets.empty();
    }
};

// Test
int main() {
    Solution solution;
    
    // Example 1: s = "()"
    cout << "Output: " << (solution.isValid("()") ? "true" : "false") << endl;
    
    // Example 2: s = "()[]{}"
    cout << "Output: " << (solution.isValid("()[]{}") ? "true" : "false") << endl;
    
    // Example 3: s = "(]"
    cout << "Output: " << (solution.isValid("(]") ? "true" : "false") << endl;
    
    // Example 4: s = "([])"
    cout << "Output: " << (solution.isValid("([])") ? "true" : "false") << endl;
    
    // Example 5: s = "([)]"
    cout << "Output: " << (solution.isValid("([)]") ? "true" : "false") << endl;

    return 0;
}
